### ISSUE TITLE
possible memory leak in ops.index_update

### ISSUE THREAD

**Issue by dchatterjee172** [google/jax#1841]
```
Line #    Mem usage    Increment   Line Contents
================================================
    98    201.1 MiB    201.1 MiB       @profile
    99                                 def _set_new_experience(self, sensor_data_one, label, reward_delta_one):
   100    201.1 MiB      0.0 MiB           min_index = jnp.argmin(self.reward_delta)
   101                                     if (
   102    201.1 MiB      0.0 MiB               jnp.abs(reward_delta_one) < self.reward_delta.min()
   103                                         and np.random.uniform(0, 1) > 0.1
   104                                     ):
   105                                         return self
   106    201.1 MiB      0.0 MiB           min_index = ops.index[min_index]
   107    201.1 MiB      0.0 MiB           sensor_data, labels, reward_delta = self
   108    201.6 MiB    ->0.5 MiB           sensor_data = ops.index_update(sensor_data, min_index, sensor_data_one)
   109    201.8 MiB    ->0.3 MiB           labels = ops.index_update(
   110    201.8 MiB      0.0 MiB               labels,
   111    201.8 MiB      0.0 MiB               min_index,
   112    201.8 MiB      0.0 MiB               label if reward_delta_one > 0 else (1 - label) * 1 / (label.size - 1),
   113                                     )
   114    201.8 MiB      0.0 MiB           reward_delta = ops.index_update(
   115    201.8 MiB      0.0 MiB               reward_delta, min_index, jnp.abs(reward_delta_one)
   116                                     )
   117    201.8 MiB      0.0 MiB           return Memory(sensor_data=sensor_data, labels=labels, reward_delta=reward_delta)


```

**Comment by dchatterjee172** [comment]
```
Line #    Mem usage    Increment   Line Contents
================================================
    98    181.5 MiB    181.5 MiB       @profile
    99                                 def _set_new_experience(self, sensor_data_one, label, reward_delta_one):
   100    181.5 MiB      0.0 MiB           min_index = np.argmin(self.reward_delta)
   101                                     if (
   102    181.5 MiB      0.0 MiB               np.abs(reward_delta_one) < self.reward_delta.min()
   103                                         and np.random.uniform(0, 1) > 0.1
   104                                     ):
   105                                         return self
   106    181.5 MiB      0.0 MiB           sensor_data, labels, reward_delta = self
   107    181.5 MiB      0.0 MiB           sensor_data[min_index, ...] = sensor_data_one
   108                                     labels[min_index, ...] = (
   109    181.5 MiB      0.0 MiB               label if reward_delta_one > 0 else (1 - label) * 1 / (label.size - 1)
   110                                     )
   111    181.5 MiB      0.0 MiB           reward_delta[min_index] = np.abs(reward_delta_one)
   112    181.5 MiB      0.0 MiB           return Memory(sensor_data=sensor_data, labels=labels, reward_delta=reward_delta)
```

same thing in numpy.

**Comment by hawkinsp** [comment]
I'm wondering if this is mostly an artifact of the instrumentation, and whether this is actually a leak or perhaps something else (for example, the JIT cache entries growing over time). I'm not familiar with the tool you used: can you give instructions to reproduce this for a small self-contained example?

**Comment by dchatterjee172** [comment]
I used https://pypi.org/project/memory-profiler/

I will try to post a snippet.

**Comment by dchatterjee172** [comment]
Is there a way to clear JIT cache entries?

**Comment by dchatterjee172** [comment]
Ami I really missing something here?

```
def worker(queue):
    rd = redis.Redis()
    flag = True

    def handle(*args):
        nonlocal flag
        flag = False

    signal(SIGTERM, handle)

    @profile
    def forward(compute: Compute, data: np.ndarray):
        o1 = jnp.matmul(data, compute.w1) + compute.b1
        a1 = jnp.tanh(o1)
        o2 = jnp.matmul(a1, compute.w2) + compute.b2
        a2 = o2 - jnp.log(jnp.exp(o2).sum(axis=1, keepdims=True))
        return a2

    @profile
    def loss(compute: Compute, data: np.ndarray, labels: np.ndarray):
        pred = forward(compute, data)
        loss = jnp.mean(-(labels * pred).sum(1))
        return loss

    g_loss = grad(loss)

    while flag:
        data = queue.get()
        id_, compute, data, labels, epoch, learning_rate = data
        grad_loss = partial(g_loss, data=data, labels=labels)

        @profile
        def sgd(
            compute: Compute, data: np.ndarray, labels: np.ndarray, learning_rate: float
        ):
            g = grad_loss(compute)
            return compute - g * learning_rate

        for i in range(1):
            if not flag:
                break
            compute = sgd(compute, data, labels, learning_rate)
        rd.set(id_, dumps(compute), ex=60)
        del data, grad_loss

```
The above is the code I am running,
```
Line #    Mem usage    Increment   Line Contents
================================================
   176    157.4 MiB    157.4 MiB       @profile
   177                                 def forward(compute: Compute, data: np.ndarray):
   178    158.2 MiB      0.8 MiB           o1 = jnp.matmul(data, compute.w1) + compute.b1
   179    160.2 MiB      2.1 MiB           a1 = jnp.tanh(o1)
   180    160.2 MiB      0.0 MiB           o2 = jnp.matmul(a1, compute.w2) + compute.b2
   181    162.0 MiB      1.8 MiB           a2 = o2 - jnp.log(jnp.exp(o2).sum(axis=1, keepdims=True))
   182    162.0 MiB      0.0 MiB           return a2


Filename: /life/components/brain.py

Line #    Mem usage    Increment   Line Contents
================================================
   184    157.4 MiB    157.4 MiB       @profile
   185                                 def loss(compute: Compute, data: np.ndarray, labels: np.ndarray):
   186    162.0 MiB      4.6 MiB           pred = forward(compute, data)
   187    162.2 MiB      0.1 MiB           loss = jnp.mean(-(labels * pred).sum(1))
   188    162.2 MiB      0.0 MiB           return loss


53
54
Filename: /life/components/brain.py

Line #    Mem usage    Increment   Line Contents
================================================
   197    157.4 MiB    157.4 MiB           @profile
   198                                     def sgd(
   199                                         compute: Compute, data: np.ndarray, labels: np.ndarray, learning_rate: float
   200                                     ):
   201    165.5 MiB      8.1 MiB               g = grad_loss(compute)
   202    165.5 MiB      0.0 MiB               return compute - g * learning_rate

```

**Comment by hawkinsp** [comment]
A good way to tease apart caching effects is: what happens to memory usage the *second* time you run the same code?

I'd really need a self-contained example I can run before I can look myself.

**Comment by dchatterjee172** [comment]
I have noticed that if the `data` and `labels` are the **SAME** in more than one calls, the memory usage does not increases.

It only happens if  `data` and `labels` are changing.

Now from the above phenomena, I am predicting that some kind of cache is building up.

**Comment by hawkinsp** [comment]
JAX tends to recompile functions each time the shapes of the inputs change. So if you feed the same shapes, if not the same values, twice, then you would expect a cache hit.

Note also that I'd recommend using `jit` if you can; it seems from a quick look that your example is not using `jit`?

**Comment by dchatterjee172** [comment]
okay, the shape is definitely changing. In every single call.
With JIT I noticed the same thing.

**Comment by dchatterjee172** [comment]
```
JAX tends to recompile functions each time the shapes of the inputs change. 
```

Why `tends`?
And is this documented somewhere?

Thanks for your help though.

**Comment by dchatterjee172** [comment]
Another question, why would recompiling the function, not free the memory?

**Comment by hawkinsp** [comment]
Recompiling doesn't free memory necessarily because `jit` keeps more than one compiled copy around. (In principle, it's a cache, although in practice I don't think we ever evict from the cache at present.)

I don't think there are any action items here (except perhaps improving the documentation), so closing this issue. But feel free to comment if you have any further questions!
